src/modules/iotjs_module_https.c

```diff
@@ -0,0 +1,862 @@
+  iotjs_https_t* https_data = (iotjs_https_t*)userp;
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_https_t, https_data);
+  if (action == CURL_POLL_IN || action == CURL_POLL_OUT ||
+      action == CURL_POLL_INOUT) {
+    iotjs_https_poll_t* poll_data = NULL;
+
+    if (!socketp) {
+      poll_data = iotjs_https_poll_create(_this->loop, sockfd, https_data);
+      curl_multi_assign(_this->curl_multi_handle, sockfd, (void*)poll_data);
+      _this->closing_handles = _this->closing_handles + 1;
+      if (_this->poll_data == NULL)
+        _this->poll_data = poll_data;
+      else
+        iotjs_https_poll_append(_this->poll_data, poll_data);
+    } else
+      poll_data = (iotjs_https_poll_t*)socketp;
+
+    if (action == CURL_POLL_IN)
+      uv_poll_start(iotjs_https_poll_get_poll_handle(poll_data), UV_READABLE,
+                    iotjs_https_uv_poll_callback);
+    else if (action == CURL_POLL_OUT)
+      uv_poll_start(iotjs_https_poll_get_poll_handle(poll_data), UV_WRITABLE,
+                    iotjs_https_uv_poll_callback);
+    else if (action == CURL_POLL_INOUT)
+      uv_poll_start(iotjs_https_poll_get_poll_handle(poll_data),
+                    UV_READABLE | UV_WRITABLE, iotjs_https_uv_poll_callback);
+  } else {
+    if (socketp) {
+      iotjs_https_poll_t* poll_data = (iotjs_https_poll_t*)socketp;
+      iotjs_https_poll_close(poll_data);
+      curl_multi_assign(_this->curl_multi_handle, sockfd, NULL);
+    }
+  }
+  return 0;
+}
+
+// Socket Assigned Callback
+int iotjs_https_curl_sockopt_callback(void* userp, curl_socket_t curlfd,
+                                      curlsocktype purpose) {
+  iotjs_https_t* https_data = (iotjs_https_t*)userp;
+  iotjs_https_jcallback(https_data, IOTJS_MAGIC_STRING_ONSOCKET,
+                        iotjs_jargs_get_empty(), false);
+  return CURL_SOCKOPT_OK;
+}
+
+// Curl wants us to signal after timeout
+int iotjs_https_curl_start_timeout_callback(CURLM* multi, long timeout_ms,
+                                            void* userp) {
+  iotjs_https_t* https_data = (iotjs_https_t*)userp;
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_https_t, https_data);
+  if (timeout_ms < 0)
+    uv_timer_stop(&(_this->timeout));
+  else {
+    if (timeout_ms == 0)
+      timeout_ms = 1;
+    if ((_this->timeout_ms != -1) && (timeout_ms > _this->timeout_ms))
+      timeout_ms = _this->timeout_ms;
+    uv_timer_start(&(_this->timeout), iotjs_https_uv_timeout_callback,
+                   (uint64_t)timeout_ms, 0);
+  }
+  return 0;
+}
+
+// Write Callback is actually to read data from incoming response
+size_t iotjs_https_curl_write_callback(void* contents, size_t size,
+                                       size_t nmemb, void* userp) {
+  iotjs_https_t* https_data = (iotjs_https_t*)userp;
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_https_t, https_data);
+  size_t real_size = size * nmemb;
+  if (iotjs_jval_is_null(&_this->jthis_native))
+    return real_size - 1;
```

Comment:
This function is a callback which `libcurl` sends us to write data (see  [https://curl.haxx.se/libcurl/c/CURLOPT_WRITEFUNCTION.html](https://curl.haxx.se/libcurl/c/CURLOPT_WRITEFUNCTION.html)). Qoute from the above link - 

> Your callback should return the number of bytes actually taken care of. If that amount differs from the amount passed to your callback function, it'll signal an error condition to the library. This will cause the transfer to get aborted.
Callback should return the number of bytes actually taken care of. If that amount differs from the amount passed to your callback function, it'll signal an error condition to the library. This will cause the transfer to get aborted.

The `real_size - 1` is used to signal an error in writing data to `libcurl`, and `real_size -1` will **always** differ from the "amount passed to your callback function" (which is `real_size`)